libs/utils/src/http/endpoint.rs

```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
+                if n > 0 {
+                    let ready = self.buffer.split().freeze();
+                    tracing::trace!(n = ready.len(), "flushing");
+
+                    // not ideal to call from blocking code to block_on, but we are sure that this
+                    // operation does not spawn_blocking other tasks
+                    let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
+                        self.tx.send(Ok(ready)).await.map_err(|_| ())?;
+
+                        // throttle sending to allow reuse of our buffer in `write`.
+                        self.tx.reserve().await.map_err(|_| ())?;
+
+                        // now the response task has picked up the buffer and hopefully started
+                        // sending it to the client.
+                        Ok(())
+                    });
+                    if res.is_err() {
+                        return Err(std::io::ErrorKind::BrokenPipe.into());
+                    }
+                    self.written += n;
+                }
+                Ok(n)
+            }
 
-    let response = Response::builder()
-        .status(200)
-        .header(CONTENT_TYPE, encoder.format_type())
-        .body(Body::from(buffer))
-        .unwrap();
+            fn flushed_bytes(&self) -> usize {
+                self.written
+            }
+        }
 
-    Ok(response)
+        impl std::io::Write for ChannelWriter {
+            fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+                let remaining = self.buffer.capacity() - self.buffer.len();
+
+                let out_of_space = remaining < buf.len();
+
+                if out_of_space {
+                    self.flush0()?;
+                }
+
+                // assume that this will often under normal operation just move the pointer back to the
+                // beginning of allocation, because previous split off parts are already sent and
+                // dropped.
+                self.buffer.extend_from_slice(buf);
+                Ok(buf.len())
+            }
+
+            fn flush(&mut self) -> std::io::Result<()> {
+                self.flush0().map(|_| ())
+            }
+        }
+
+        let mut writer = ChannelWriter::new(128 * 1024, tx);
+
+        let encoder = TextEncoder::new();
+
+        let response = Response::builder()
+            .status(200)
+            .header(CONTENT_TYPE, encoder.format_type())
+            .body(body)
+            .unwrap();
+
+        let span = info_span!("blocking");
+
+        tokio::task::spawn_blocking(move || {
+            let _span = span.entered();
+            // Currently we take a lot of mutexes while collecting metrics, so it's
+            // better to spawn a blocking task to avoid blocking the event loop.
+            let metrics = metrics::gather();
+            let res = encoder
+                .encode(&metrics, &mut writer)
+                .and_then(|_| writer.flush().map_err(|e| e.into()));
```

Comment:
Would be helpful to annotate the `e.into()`. Reading top down, especially with the elaborate `Err(e)` below, it's hard to follow what error type this is.

I _think_ we're converting encoder error into io error? https://docs.rs/prometheus/latest/prometheus/enum.Error.html

Reply:
Yeah this was hugely frustrating. I hope we could get rid of prometheus asap.

Reply:
This transform is made *because* TextEncoder can fail for non-IO reasons, which we handle in the most confusing `blocking_send(Err(BrokenPipe.into()))` for which I don't know a better alternative (we want to signal to hyper that something went bad, drop the connection).